dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/Pickler.cs

```diff
@@ -1,773 +1,128 @@
-			put_set((IEnumerable)o);
-			return true;
-		}
-
-		var dictionary = o as IDictionary;
-		if(dictionary != null) {
-			put_map(dictionary);
-			return true;
-		}
-
-		var list = o as IList;
-		if(list != null) {
-			put_enumerable(list);
-			return true;
-		}
-
-		var enumerable = o as IEnumerable;
-		if(enumerable != null) {
-			put_enumerable(enumerable);
-			return true;
-		}
-		
-		DataContractAttribute dca = (DataContractAttribute) Attribute.GetCustomAttribute(t, typeof(DataContractAttribute));
-		if(dca!=null) {
-			put_datacontract(t, o, dca);
-			return true;
-		}
-		
-		SerializableAttribute sa = (SerializableAttribute) Attribute.GetCustomAttribute(t, typeof(SerializableAttribute));
-		if(sa!=null) {
-			put_serializable(t, o);
-			return true;
-		}
-		
-		if(hasPublicProperties(o)) {
-			put_objwithproperties(o);
-			return true;
-		}
-
-		return false;
-	}
-	
-	[SuppressMessage("ReSharper", "MemberCanBeMadeStatic.Global")]
-	protected IObjectPickler getCustomPickler(Type t)
-	{
-		IObjectPickler pickler;
-		if(customPicklers.TryGetValue(t, out pickler))
-			return pickler;		// exact match
-		
-		// check if there's a custom pickler registered for an interface or abstract base class
-		// that this object implements or inherits from.
-		foreach(var x in customPicklers) {
-			if(x.Key.IsAssignableFrom(t)) {
-				return x.Value;
-			}
-		}
-
-		return null;
-	}
-
-	private static bool hasPublicProperties(object o)
-	{
-		var props=o.GetType().GetProperties();
-		return props.Length>0;
-	}
-
-	private void put_datetime(DateTime dt) {
-		outs.WriteByte(Opcodes.GLOBAL);
-		outs.Write(datetimeDatetimeBytes, 0, datetimeDatetimeBytes.Length);
-		outs.WriteByte(Opcodes.MARK);
-        put_long(dt.Year);
-        put_long(dt.Month);
-        put_long(dt.Day);
-        put_long(dt.Hour);
-        put_long(dt.Minute);
-        put_long(dt.Second);
-        put_long(dt.Millisecond * 1000);
-		outs.WriteByte(Opcodes.TUPLE);
-		outs.WriteByte(Opcodes.REDUCE);
-        WriteMemo(dt);
-	}
-
-	private void put_timespan(TimeSpan ts) {
-		outs.WriteByte(Opcodes.GLOBAL);
-		outs.Write(datetimeTimedeltaBytes, 0, datetimeTimedeltaBytes.Length);
-        put_long(ts.Days);
-		put_long(ts.Hours*3600+ts.Minutes*60+ts.Seconds);
-        put_long(ts.Milliseconds*1000);
-		outs.WriteByte(Opcodes.TUPLE3);
-		outs.WriteByte(Opcodes.REDUCE);
-        WriteMemo(ts);
-	}
-
-	private void put_enumerable(IEnumerable list) {
-		outs.WriteByte(Opcodes.EMPTY_LIST);
-		WriteMemo(list);
-		outs.WriteByte(Opcodes.MARK);
-		foreach(var o in list) {
-			save(o);
-		}
-		outs.WriteByte(Opcodes.APPENDS);
-	}
-
-	private void put_map(IDictionary o) {
-		outs.WriteByte(Opcodes.EMPTY_DICT);
-		WriteMemo(o);
-		outs.WriteByte(Opcodes.MARK);
-		foreach(var k in o.Keys) {
-			save(k);
-			save(o[k]);
-		}
-		outs.WriteByte(Opcodes.SETITEMS);
-	}
+            _picklerImplementation = new PicklerImplementation<ArrayWriter>(new ArrayWriter(new byte[64]), this, useMemo);
```

Comment:
If someone calls this method multiple times, we will create a new PicklerImplementation each time, but we don't dispose the old one. Is this intended?

Reply:
The last remark makes sense, there's no real need to have the pickler be IDisposable. However changing this will break existing code

Reply:
Very good point. I added a call to `close()` method to make sure the previous stream gets disposed.